### PR TITLE
fix: Add deployment update path

### PR DESCRIPTION
**Pull request by wob-adarga**
The upstream pipeline-controller allows for a gitops style approach, updating manifests cause new deployments; rather than use pachctl from client box to effect this. This allows us to integrate nicely to argocd.

The upstream version currently lacks support for amazon storage and has some slightly different deployment characteristics when comparing with pods deployed in the current PoC env, ds-lab-1. Going through these differences and attempting to remove them by way of code changes in this downstream repo. Not bug fixes, so not appropriate for PR into upstream, and also lacking any sort of config so very incomplete, just more PoC....

(Annotated changes with Adarga comment, without knowing exactly how we will currently play with upstream, may help bringing in / reconciling changes from pachyderm end.)

APP-2504

### CODE REVIEW

File: controllers/pipeline_controller.go [hunk 1]
```diff
@@ -81,7 +82,28 @@ func (r *PipelineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	} else if err != nil {
 		return ctrl.Result{}, err
+		// Adarga added: update the deploymnent in case of delta
+		// wob: 2021-02-22
+	} else {
```
Comment (ollystephens) [inline controllers/pipeline_controller.go:87]: This shouldn't be in an else clause.

File: controllers/pipeline_controller.go [hunk 2]
```diff
@@ -81,7 +82,28 @@ func (r *PipelineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		if err := controllerutil.SetControllerReference(&pipeline, pipelineDeployment, r.Scheme); err != nil {
```
Comment (ollystephens) [inline controllers/pipeline_controller.go:88]: can re-use existing err var here

File: controllers/pipeline_controller.go [hunk 3]
```diff
@@ -81,7 +82,28 @@ func (r *PipelineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		log.Info("Updating Deployment: ", pipelineDeployment.Namespace, pipelineDeployment.Name)
+
+		err = r.Update(ctx, pipelineDeployment)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		pipeline.Status.State = "Running"
+
+		if err := r.Status().Update(ctx, &pipeline); err != nil {
```
Comment (ollystephens) [inline controllers/pipeline_controller.go:101]: and here

**Review by ollystephens — APPROVED** [review]
Couple of minor comments; not worth holding up the merge though unless you want to fix them first.

**Review by saminahbab — APPROVED** [review]

File: controllers/pipeline_controller.go (hunk 1 above)
Comment (wob-adarga) [reply controllers/pipeline_controller.go:87]: Yep gofmt was adamant about that too, I really should have listened.

**Comment by wob-adarga** [comment]
> Couple of minor comments; not worth holding up the merge though unless you want to fix them first.

Thanks Olly, I'll incorporate into the inevitable next commit.

File: controllers/pipeline_controller.go (hunk 1 above)
Comment (wob-adarga) [reply controllers/pipeline_controller.go:87]: Actually, are you talking about the comment or the main update bit. The update bit needs to be in else to ensure the deploy was pre-existing ...

**Pull request merged**